src/somnium/congomongo.clj

```diff
@@ -110,23 +110,38 @@
   "Makes a connection with passed database name, [{:host host, :port port}]
   server addresses and MongoClientOptions.
 
-  username and password may be supplied for authenticated connections."
-  [db {:keys [instances options username password]}]
+  username, password, and optionally auth-source, may be supplied for authenticated connections."
+  [db {:keys [instances options username password] {auth-mechanism :mechanism auth-source :source} :auth-source}]
   (when (not= (nil? username) (nil? password))
     (throw (IllegalArgumentException. "Username and password must both be supplied for authenticated connections")))
 
   (let [addresses (map (fn [{:keys [host port] :or {host "127.0.0.1" port 27017}}]
-                        (make-server-address host port))
+                         (make-server-address host port))
                        (or instances [{:host "127.0.0.1" :port 27017}]))
 
         ^MongoClientOptions options (or options (mongo-options))
+        ^MongoCredential credential (cond
+                                      (and username password (= :plain auth-mechanism))
+                                      (MongoCredential/createPlainCredential username auth-source
+                                                                             (.toCharArray password))
 
-        mongo (cond
-                (and username password) (make-mongo-client
-                                          addresses
-                                          [(MongoCredential/createCredential username db (.toCharArray password))]
-                                          options)
-                :else (make-mongo-client addresses options))
+                                      (and username password (= :scram-1 auth-mechanism))
+                                      (MongoCredential/createScramSha1Credential username auth-source
+                                                                                 (.toCharArray password))
+
+                                      (and username password (= :scram-256 auth-mechanism))
+                                      (MongoCredential/createScramSha256Credential username auth-source
+                                                                                   (.toCharArray password))
+
+                                      auth-mechanism
+                                      (throw (UnsupportedOperationException. (str auth-mechanism " is not supported.")))
```

Comment:
If username or password is `nil` and a valid auth-mechanism is provided, would this claim that mechanism isn't supported?

Reply:
In that scenario it's a completely invalid configuration but I'll tweak it to remove the ambiguity,